### PR TITLE
Flux Monitor uses a single log subscription for all FluxAggregators (WIP)

### PR DESCRIPTION
**Pull request by spooktheducks**

### CODE REVIEW

File: core/eth/log_broadcaster.go [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+	b.chAddListener <- registration{address, listener}
+}
+
+func (b *logBroadcaster) Unregister(address common.Address, listener LogListener) {
+	b.chRemoveListener <- registration{address, listener}
+}
+
+// Subscriptions are closed and then recreated when the set of contracts we're listening
+// to changes.
+func (b *logBroadcaster) resubscribe() {
```
Comment (samsondav) [inline core/eth/log_broadcaster.go]: start resubscribe loop ?

File: core/eth/log_broadcaster.go [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+			select {
+			case <-b.chStop:
+				return
+			default:
+				time.Sleep(10 * time.Second) // A configurable timeout might be useful here
+				continue ResubscribeLoop
+			}
+		}
+
+		shouldResubscribe := b.process(subscription, chRawLogs)
```
Comment (samsondav) [inline core/eth/log_broadcaster.go]: can we call it stop or something?

File: core/eth/log_broadcaster_test.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+	var recvd []interface{}
+	lb.Register(addr, eth.NewFuncLogListener(func(log interface{}, err error) {
+		require.NoError(t, err)
+		recvd = append(recvd, log)
+	}))
+
+	chRawLogs := <-chchRawLogs
+
+	for i := 0; i < len(logs); i++ {
```
Comment (spooktheducks) [inline core/eth/log_broadcaster_test.go]: Add comment: `// Simulates resuming the subscription repeatedly`

File: core/eth/log_broadcaster_test.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+	ethClient.On("SubscribeToLogs", mock.Anything, mock.Anything).
+		Run(func(args mock.Arguments) {
+			chchRawLogs <- args.Get(0).(chan<- eth.Log)
+		}).
+		Return(sub, nil).
+		Once()
+	ethClient.On("SubscribeToLogs", mock.Anything, mock.Anything).
+		Run(func(args mock.Arguments) {
+			query := args.Get(1).(ethereum.FilterQuery)
+			require.Equal(t, big.NewInt(expectedBlock), query.FromBlock)
```
Comment (spooktheducks) [inline core/eth/log_broadcaster_test.go]: Need to figure out how to test the contents of `query.Addresses` despite the fact that its contents are randomly ordered.

File: core/services/fluxmonitor/flux_monitor.go [hunk 5]
```diff
@@ -378,284 +328,210 @@ func NewPollingDeviationChecker(
...
+		case <-p.chStop:
 			return
-		case maybeLog := <-roundSubscription.Logs():
-			if maybeLog.Error != nil {
-				logger.Error(errors.WithStack(maybeLog.Error))
+		case maybeLog := <-p.chMaybeLogs:
+			if maybeLog.Err != nil {
+				logger.Error(errors.WithStack(maybeLog.Err))
+				// @@TODO: other error handling?
+				// @@TODO: detect broken subscription?
```
Comment (spooktheducks) [inline core/services/fluxmonitor/flux_monitor.go]: Actually, subscriptions are handled in LogSubscriber now, so errors here would be from the DecodingLogListener, if anything.

File: core/services/fluxmonitor/flux_monitor.go [hunk 6]
```diff
@@ -187,58 +157,18 @@ func (fm *concreteFluxMonitor) AddJob(job models.JobSpec) error {
...
-			}
-		}
+	if len(validCheckers) == 0 {
+		return nil
 	}
 
-	if len(checkers) > 0 {
-		jobMap[jobSpecID] = checkers
-	}
+	fm.adds <- addEntry{*job.ID, validCheckers}
```
Comment (samsondav) [inline core/services/fluxmonitor/flux_monitor.go]: Dereference nil check?

File: core/services/fluxmonitor/flux_monitor.go [hunk 7]
```diff
@@ -187,58 +157,18 @@ func (fm *concreteFluxMonitor) AddJob(job models.JobSpec) error {
...
 	return nil
 }
 
 // RemoveJob stops and removes the checker for all Flux Monitor initiators belonging
 // to the passed job ID.
-func (fm *concreteFluxMonitor) RemoveJob(ID *models.ID) {
-	fm.removes <- ID
+func (fm *concreteFluxMonitor) RemoveJob(id *models.ID) {
+	fm.removes <- *id
```
Comment (samsondav) [inline core/services/fluxmonitor/flux_monitor.go]: and here

File: core/services/fluxmonitor/flux_monitor.go [hunk 8]
```diff
@@ -187,58 +157,18 @@ func (fm *concreteFluxMonitor) AddJob(job models.JobSpec) error {
...
-	}
+	fm.adds <- addEntry{*job.ID, validCheckers}
 	return nil
 }
 
 // RemoveJob stops and removes the checker for all Flux Monitor initiators belonging
 // to the passed job ID.
-func (fm *concreteFluxMonitor) RemoveJob(ID *models.ID) {
-	fm.removes <- ID
+func (fm *concreteFluxMonitor) RemoveJob(id *models.ID) {
```
Comment (samsondav) [inline core/services/fluxmonitor/flux_monitor.go]: pass value instead of pointer?

File: core/services/fluxmonitor/flux_monitor.go [hunk 9]
```diff
@@ -334,40 +265,59 @@ func GetBridgeURLFromName(name string, orm *orm.ORM) (*url.URL, error) {
...
+	chStop     chan struct{}
 	waitOnStop chan struct{}
 }
 
+type maybeLog struct {
+	Log interface{}
+	Err error
+}
+
+type AtomicBool struct {
```
Comment (samsondav) [inline core/services/fluxmonitor/flux_monitor.go]: move this?

File: core/services/fluxmonitor/flux_monitor.go [hunk 10]
```diff
@@ -378,284 +328,210 @@ func NewPollingDeviationChecker(
...
-// the deviation against.
-func (p *PollingDeviationChecker) fetchAggregatorData() error {
-	price, err := p.fluxAggregator.LatestAnswer(p.precision)
-	if err != nil {
-		return err
+	// It's pointless to listen to logs from before the current reporting round
+	p.reportableRoundID = roundState.ReportableRoundID
+
+	if !roundState.EligibleToSubmit {
+		logger.Info("not eligible to submit, polling paused")
```
Comment (samsondav) [inline core/services/fluxmonitor/flux_monitor.go]: Does it really pause the polling? Or just skip doing anything?

File: evm-contracts/src/v0.6/dev/FluxAggregator.sol [hunk 11]
```diff
@@ -674,6 +674,42 @@ contract FluxAggregator is AggregatorInterface, Owned {
...
+    view
+    returns (uint32 _reportableRoundId, bool _eligibleToSubmit, int256 _latestAnswer)
+  {
+    if (finished(reportingRoundId) || timedOut(reportingRoundId)) {
+      return (reportingRoundId.add(1), eligibleToSubmit(), rounds[latestRoundId].answer);
+    }
+    return (reportingRoundId, eligibleToSubmit(), rounds[latestRoundId].answer);
+  }
+
+  function eligibleToSubmit()
```
Comment (samsondav) [inline evm-contracts/src/v0.6/dev/FluxAggregator.sol]: update logic... don't pass _roundId

File: core/services/fluxmonitor/flux_monitor.go [hunk 12]
```diff
@@ -378,284 +328,210 @@ func NewPollingDeviationChecker(
...
-	nextPrice, err := p.fetchPrices()
-	if err != nil {
-		return false, err
-	}
-	promSetDecimal(promFMSeenValue.WithLabelValues(jobSpecID), nextPrice)
-
-	if !OutsideDeviation(p.currentPrice, nextPrice, threshold) {
-		return false, nil // early exit since deviation criteria not met.
+	if !OutsideDeviation(utils.DecimalFromBigInt(roundState.LatestAnswer, p.precision), nextAnswer, threshold) {
+		return
```
Comment (samsondav) [inline core/services/fluxmonitor/flux_monitor.go]: log debug here?

File: core/services/fluxmonitor/flux_monitor.go [hunk 13]
```diff
@@ -378,284 +328,210 @@ func NewPollingDeviationChecker(
...
 	}
 
-	nextRound := new(big.Int).Add(p.currentRound, big.NewInt(1)) // start new round
 	logger.Infow("Detected change outside threshold, starting new round",
-		"round", nextRound,
+		"reporatableRound", roundState.ReportableRoundID,
```
Comment (samsondav) [inline core/services/fluxmonitor/flux_monitor.go]: reportable*

File: core/services/fluxmonitor/flux_monitor.go [hunk 14]
```diff
@@ -378,284 +328,210 @@ func NewPollingDeviationChecker(
...
+	p.reportableRoundID = roundState.ReportableRoundID
+
+	if !roundState.EligibleToSubmit {
+		logger.Info("not eligible to submit, polling paused")
+		return
 	}
-	p.currentPrice = price
 
-	round, err := p.fluxAggregator.LatestRound()
+	nextAnswer, err := p.fetcher.Fetch()
```
Comment (samsondav) [inline core/services/fluxmonitor/flux_monitor.go]: polledAnswer?

File: core/services/fluxmonitor/flux_monitor_test.go [hunk 15]
```diff
@@ -36,12 +37,12 @@ func (f *successFetcher) Fetch() (decimal.Decimal, error) {
...
-	sub.On("Unsubscribe").Return()
-	sub.On("Logs").Return(make(<-chan contracts.MaybeDecodedLog))
-	return sub
-}
+//func fakeSubscription() *mocks.LogSubscription {
+//    sub := new(mocks.LogSubscription)
+//    sub.On("Unsubscribe").Return()
+//    sub.On("Logs").Return(make(<-chan contracts.MaybeDecodedLog))
+//    return sub
+//}
```
Comment (samsondav) [inline core/services/fluxmonitor/flux_monitor_test.go]: fun

File: core/services/fluxmonitor/flux_monitor_test.go [hunk 16]
```diff
@@ -186,291 +163,387 @@ func TestConcreteFluxMonitor_StopWithoutStart(t *testing.T) {
...
 
-	deviationChecker.Stop()
+			checker.ExportedPollIfEligible(test.threshold)
 
-	assert.Equal(t, decimal.NewFromInt(100).String(), deviationChecker.ExportedCurrentPrice().String())
+			fluxAggregator.AssertExpectations(t)
+			fetcher.AssertExpectations(t)
+			rm.AssertExpectations(t)
+		})
+	}
```
Comment (samsondav) [inline core/services/fluxmonitor/flux_monitor_test.go]: :+1:

File: core/services/fluxmonitor/flux_monitor_test.go [hunk 17]
```diff
@@ -186,291 +163,387 @@ func TestConcreteFluxMonitor_StopWithoutStart(t *testing.T) {
...
+			fluxAggregator := new(mocks.FluxAggregator)
+			fetcher := new(mocks.Fetcher)
 
-func TestPollingDeviationChecker_StartStop(t *testing.T) {
-	store, cleanup := cltest.NewStore(t)
-	defer cleanup()
+			job := cltest.NewJobWithFluxMonitorInitiator()
+			initr := job.Initiators[0]
+			initr.ID = 1
+			initr.PollingInterval = models.Duration(math.MaxInt64)
```
Comment (samsondav) [inline core/services/fluxmonitor/flux_monitor_test.go]: What happens if someone ever passes 0 in as a PollingInterval? Probably want to validate this.

File: core/utils/utils.go [hunk 18]
```diff
@@ -515,3 +516,11 @@ func Uint256ToHex(n *big.Int) (string, error) {
 	}
 	return common.BigToHash(n).Hex(), nil
 }
+
+var dec10 = decimal.NewFromInt(10)
+
+func DecimalFromBigInt(i *big.Int, precision int32) decimal.Decimal {
+	raw := decimal.NewFromBigInt(i, 0)
```
Comment (samsondav) [inline core/utils/utils.go]: are we gonna change this?

**Comment by samsondav** [comment]
And a bit more debug logging probably wouldn't hurt.

File: core/services/fluxmonitor/flux_monitor.go [hunk 19]
```diff
@@ -78,104 +74,78 @@ func New(
...
+		adds:       make(chan addEntry),
+		removes:    make(chan models.ID),
+		connect:    make(chan *models.Head),
+		disconnect: make(chan struct{}),
+		chStop:     make(chan struct{}),
+		chDone:     make(chan struct{}),
 	}
 }
 
 func (fm *concreteFluxMonitor) Start() error {
```
[the remark this answers is not in the document]
Comment (spooktheducks) [reply core/services/fluxmonitor/flux_monitor.go]: Will address this in a separate PR.

File: core/services/fluxmonitor/flux_monitor.go (hunk 8 above)
Comment (spooktheducks) [reply core/services/fluxmonitor/flux_monitor.go]: Would break the interface

File: core/services/fluxmonitor/flux_monitor_test.go (hunk 17 above)
Comment (spooktheducks) [reply core/services/fluxmonitor/flux_monitor_test.go]: Handled in `pollingDeviationCheckerFactory#New`

File: core/utils/utils.go (hunk 18 above)
Comment (felder-cl) [reply core/utils/utils.go]: "are we gonna change this?" how come?  what's the context?

Comment (felder-cl) [reply core/utils/utils.go]: Could we not use decimal.NewFromBigInt(..) instead? https://godoc.org/github.com/shopspring/decimal#NewFromBigInt

**Comment by felder-cl** [comment]
This PR looks like its still against the flux-aggregator-wrapper branch.  Maybe change it to develop?

File: core/eth/contracts.go [hunk 20]
```diff
@@ -63,8 +63,8 @@ func GetV6ContractCodec(name string) (ContractCodec, error) {
 	return getContractCodec(name, box)
 }
 
-func (contract *contractCodec) ABI() abi.ABI {
-	return contract.abi
+func (contract *contractCodec) ABI() *abi.ABI {
+	return &contract.abi
 }
```
Comment (felder-cl) [inline core/eth/contracts.go]: Should this return a copy of the ABI to try and prevent altering internal structure, or what that the point>

File: core/eth/contracts.go [hunk 21]
```diff
@@ -113,16 +113,25 @@ func (cc *contractCodec) UnpackLog(out interface{}, event string, log Log) error
...
+type UnsubscribeFunc func()
+
+func NewConnectedContract(
+	codec ContractCodec,
+	address common.Address,
+	ethClient Client,
+	logBroadcaster LogBroadcaster,
+) ConnectedContract {
+	return &connectedContract{codec, address, ethClient, logBroadcaster}
 }
```
Comment (felder-cl) [inline core/eth/contracts.go]: what's a Log Broadcaster?  Why does a connected contract have to know about it?

File: core/eth/contracts.go [hunk 22]
```diff
@@ -113,16 +113,25 @@ func (cc *contractCodec) UnpackLog(out interface{}, event string, log Log) error
...
+	ethClient      Client
+	logBroadcaster LogBroadcaster
 }
 
-func NewConnectedContract(cc ContractCodec, ethClient Client, address common.Address) ConnectedContract {
-	return &connectedContract{cc, ethClient, address}
+type UnsubscribeFunc func()
+
+func NewConnectedContract(
+	codec ContractCodec,
```
Comment (felder-cl) [inline core/eth/contracts.go]: where is this codec used?

**Pull request closed without merging**